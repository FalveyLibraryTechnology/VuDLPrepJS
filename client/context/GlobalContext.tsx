import React, { createContext, useContext, useReducer, Dispatch, ReactNode } from "react";

interface SnackbarState {
    open: boolean,
    message: string,
    severity: string
}

interface GlobalState {
    // Modal control
    modalOpenStates: Record<string, boolean>;
    // Snackbar
    snackbarState: SnackbarState;
}

/**
 * Pass a shared entity to react components,
 * specifically a way to make api requests.
 */
const initalGlobalState: GlobalState = {
    // Modal control
    modalOpenStates: {},
    // Snackbar
    snackbarState: {
        open: false,
        message: "",
        severity: "info",
    },
};

type Action =
    | { type: 'OPEN_MODAL'; payload: string }
    | { type: 'CLOSE_MODAL'; payload: string }
    | { type: 'SET_SNACKBAR_STATE'; payload: SnackbarState };

/**
 * Update the shared states of react components.
 */
const globalReducer = (state: GlobalState, action: Action): GlobalState => {
    switch (action.type) {
        case 'OPEN_MODAL':
            return {
                ...state,
                modalOpenStates: {
                    ...state.modalOpenStates,
                    [action.payload]: true,
                },
            };
        case 'CLOSE_MODAL':
            return {
                ...state,
                modalOpenStates: {
                    ...state.modalOpenStates,
                    [action.payload]: false,
                },
            };
        case 'SET_SNACKBAR_STATE':
            return {
                ...state,
                snackbarState: action.payload,
            };
        default:
            return state;
    }
};

/**
 * Context to provide global state and dispatch function.
 */
interface GlobalContextProps {
    state: GlobalState;
    dispatch: Dispatch<Action>;
}

const GlobalContext = createContext<GlobalContextProps | undefined>(undefined);

/**
 * GlobalContextProvider to wrap around the application.
 */
export const GlobalContextProvider = ({ children }: { children: ReactNode }) => {
    const [state, dispatch] = useReducer(globalReducer, initalGlobalState);
    return (
        <GlobalContext.Provider value={{ state, dispatch }}>
            {children}
        </GlobalContext.Provider>
    );
};

export const useGlobalContext = () => {
    const context = useContext(GlobalContext);

    if (!context) {
        throw new Error("useGlobalContext must be used within a GlobalContextProvider");
    }

    const { state, dispatch } = context;

    // Modal control

    const isModalOpen = (modal: string) => state.modalOpenStates[modal] ?? false;
    const openModal = (modal: string) => {
        dispatch({
            type: "OPEN_MODAL",
            payload: modal
        });
    };
    const closeModal = (modal: string) => {
        dispatch({
            type: "CLOSE_MODAL",
            payload: modal
        });
    };
    const toggleModal = (modal: string) => {
        if (isModalOpen(modal)) {
            closeModal(modal);
        } else {
            openModal(modal);
        }
    };

    // Snackbar

    const setSnackbarState = (snackbarState: SnackbarState) => {
        dispatch({
            type: "SET_SNACKBAR_STATE",
            payload: snackbarState
        });
    };

    return {
        state: {
            // only return limited state
            // to create "private" attributes
            snackbarState: state.snackbarState,
        },
        action: {
            // Modal control
            isModalOpen,
            openModal,
            closeModal,
            toggleModal,
            // Snackbar
            setSnackbarState,
        },
    };
}

export default {
    GlobalContextProvider,
    useGlobalContext
}
