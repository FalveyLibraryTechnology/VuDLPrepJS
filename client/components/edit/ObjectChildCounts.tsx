import React from "react";
import { ChildCounts, useEditorContext } from "../../context/EditorContext";
import ObjectLoader from "./ObjectLoader";

export interface ObjectChildCountsProps {
    pid: string;
}

export const ObjectChildCounts = ({ pid }: ObjectChildCountsProps): React.ReactElement => {
    const {
        state: { childCountsStorage },
        action: { loadChildCountsIntoStorage },
    } = useEditorContext();
    const loaded: boolean = Object.prototype.hasOwnProperty.call(childCountsStorage, pid);
    if (!loaded) {
        loadChildCountsIntoStorage(pid);
    }
    const details: ChildCounts = loaded ? childCountsStorage[pid] : { directChildren: 0, totalDescendants: 0 };

    const stateMsg = loaded ? (
        <div>{details.directChildren} children, {details.totalDescendants} descendants</div>
    ) : (
        ""
    );
    return (
        <>
            <ObjectLoader pid={pid} />
            {stateMsg}
        </>
    );
};

export default ObjectChildCounts;
