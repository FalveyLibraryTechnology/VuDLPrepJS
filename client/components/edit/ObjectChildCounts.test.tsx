import React from "react";
import { beforeEach, describe, expect, it, jest } from "@jest/globals";
import { waitFor } from "@testing-library/react";
import renderer from "react-test-renderer";
import { ObjectChildCountsProps, ObjectChildCounts } from "./ObjectChildCounts";
import { EditorContextProvider, ChildCounts } from "../../context/EditorContext";
import { FetchContextProvider } from "../../context/FetchContext";
import { GlobalContextProvider } from "../../context/GlobalContext";

jest.mock("./ObjectLoader", () => (args) => JSON.stringify(args));

function getMountedObjectChildCountsComponent(props: ObjectChildCountsProps) {
    return renderer.create(
        <GlobalContextProvider>
            <FetchContextProvider>
                <EditorContextProvider>
                    <ObjectChildCounts {...props} />
                </EditorContextProvider>
            </FetchContextProvider>
            ,
        </GlobalContextProvider>,
    );
}

describe("ObjectChildCounts", () => {
    let props: ObjectChildCountsProps;
    let lastRequestUrl: string;
    let response: ChildCounts;

    beforeEach(() => {
        props = { pid: "foo:123" };
        response = { directChildren: 5, totalDescendants: 100 };
        global.fetch = jest.fn((url) => {
            lastRequestUrl = url as string;
            return {
                ok: true,
                status: 200,
                json: async function () {
                    return response;
                },
            };
        });
    });

    it("displays the data found in the response", async () => {
        let tree;
        await renderer.act(async () => {
            tree = getMountedObjectChildCountsComponent(props);
            await waitFor(() => expect(global.fetch).toHaveBeenCalledTimes(1));
        });
        expect(lastRequestUrl).toEqual("http://localhost:9000/api/edit/object/foo%3A123/childCounts");
        expect(tree.toJSON()).toMatchSnapshot();
    });
});
