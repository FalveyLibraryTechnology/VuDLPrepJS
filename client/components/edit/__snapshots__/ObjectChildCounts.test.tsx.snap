// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectChildCounts displays the data found in the response 1`] = `
[
  "{"pid":"foo:123"}",
  <div>
    5
     children, 
    100
     descendants
  </div>,
  ",",
]
`;
